### PR TITLE
feat(wasm): support for wasm

### PR DESCRIPTION
**Pull request by mitsuhiko**
This implements WASM supposed based on the notes in #290.

Things of note:

* WASM has no debug/code id, so I made something up for testing (https://github.com/WebAssembly/tool-conventions/issues/133)
* It uses a non unified addressing system so data and functions are separate. This might be irrelevant for us but I need to verify this
* there is no load address
* unclear about alignment values and pointer sizes. I just assumed this is all 4 bytes
* it changes the `load_address` to an `address_offset` that can be negative to handle the code section offset in WASM.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
Notes on how to approximate a symbol table: https://github.com/WebAssembly/tool-conventions/issues/38

File: symbolic-debuginfo/src/wasm.rs [hunk 1]
```diff
@@ -113,16 +110,16 @@ impl<'d> WasmObject<'d> {
     }
 
     /// Returns an iterator over symbols in the public symbol table.
-    pub fn symbols<'o>(&'o self) -> WasmSymbolIterator<'d, 'o> {
+    pub fn symbols(&self) -> WasmSymbolIterator<'data, '_> {
+        let iterator = Box::new(self.wasm_module.funcs.iter()) as Box<dyn Iterator<Item = _>>;
```
Comment (Swatinem) [inline symbolic-debuginfo/src/wasm.rs:114]: is it possible to un-box this?

Comment (mitsuhiko) [reply symbolic-debuginfo/src/wasm.rs:114]: I'm not sure. Walrus returns an `impl Iterator`.

File: symbolic-debuginfo/src/wasm.rs [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+    }
+}
+
+impl<'d> Dwarf<'d> for WasmObject<'d> {
+    fn endianity(&self) -> Endian {
+        Endian::Little
+    }
+
+    fn raw_section(&self, section_name: &str) -> Option<DwarfSection<'d>> {
+        for (_, section) in self.wasm_module.customs.iter() {
```
Comment (mitsuhiko) [inline symbolic-debuginfo/src/wasm.rs:262]: This does not cover non custom sections.

File: symbolic-debuginfo/src/dwarf.rs [hunk 3]
```diff
@@ -43,6 +43,10 @@ type IncompleteLineNumberProgram<'a> = gimli::read::IncompleteLineProgram<Slice<
 type LineNumberProgramHeader<'a> = gimli::read::LineProgramHeader<Slice<'a>>;
 type LineProgramFileEntry<'a> = gimli::read::FileEntry<Slice<'a>>;
 
+fn offset(addr: u64, offset: i64) -> u64 {
+    (addr as i128).saturating_sub(offset as i128) as u64
```
Comment (Swatinem) [inline symbolic-debuginfo/src/dwarf.rs:47]: While the `saturating_sub` is strictly better, can you explain the reasoning for this change?

File: symbolic-debuginfo/src/wasm.rs [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    }
+
+    /// The code identifier of this object.
+    ///
+    /// Wasm does not yet provide code IDs.
+    pub fn code_id(&self) -> Option<CodeId> {
+        // see `debug_id`
+        for (_, section) in self.wasm_module.customs.iter() {
+            if section.name() == "build_id" {
+                return Some(CodeId::from_binary(&section.data(&Default::default())));
```
Comment (Swatinem) [inline symbolic-debuginfo/src/wasm.rs:74]: can you extract the loop and returning the `section.data` into its own function, also, what does the `Default` expand to here? Its a bit hard to follow without actually jumping to the definition.

File: symbolic-debuginfo/src/wasm.rs [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+            }
+        }
+        false
+    }
+
+    /// Constructs a debugging session.
+    pub fn debug_session(&self) -> Result<DwarfDebugSession<'data>, DwarfError> {
+        let symbols = self.symbol_map();
+        // WASM is offset by the negative offset to the code section instead of the load address
+        DwarfDebugSession::parse(self, symbols, -(self.code_offset() as i64), self.kind())
```
Comment (Swatinem) [inline symbolic-debuginfo/src/wasm.rs:156]: the comment makes sense, but considering you are using `saturating_sub` in the other place, I wonder if an example can make all this clearer?

File: symbolic-debuginfo/src/wasm.rs [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+        self.has_unwind_info()
+    }
+
+    fn has_sources(&self) -> bool {
+        self.has_sources()
+    }
+}
+
+impl<'d> Dwarf<'d> for WasmObject<'d> {
+    fn endianity(&self) -> Endian {
```
Comment (Swatinem) [inline symbolic-debuginfo/src/wasm.rs:279]: lol, only noticing this now, but isnt `endianness` the more common word for this?

File: symbolic-debuginfo/src/wasm.rs [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+        Endian::Little
+    }
+
+    fn raw_section(&self, section_name: &str) -> Option<DwarfSection<'d>> {
+        for (_, section) in self.wasm_module.customs.iter() {
+            if section.name().starts_with('.') && &section.name()[1..] == section_name {
```
Comment (Swatinem) [inline symbolic-debuginfo/src/wasm.rs:285]: maybe use `strip_prefix`, I wonder why clippy is not complaining

**Review by Swatinem — COMMENTED** [review]
can you strip the wasm file somehow? 1.4M is quite a lot for the fixture…

File: symbolic-debuginfo/src/wasm.rs (hunk 5 above)
Comment (mitsuhiko) [reply symbolic-debuginfo/src/wasm.rs:156]: I'm changing the saturating to wrapping which makes more sense to the rest of the system.

File: symbolic-debuginfo/src/wasm.rs (hunk 6 above)
Comment (mitsuhiko) [reply symbolic-debuginfo/src/wasm.rs:279]: Blame @jan-auer

**Comment by mitsuhiko** [comment]
I compiled the example WASM file with `no_std`. Beautiful 4K now :)

**Review by Swatinem — APPROVED** [review]
